lib/rexml/parsers/baseparser.rb

```diff
@@ -460,15 +461,24 @@ def pull_event
                 @closed = tag
                 @nsstack.shift
               else
+                if @tags.size.zero? && !@root_tag.nil?
+                  raise ParseException.new("Malformed XML: Extra tag at the end of the document (got '<#{tag}')", @source)
+                end
                 @tags.push( tag )
               end
+              @root_tag ||= tag
               return [ :start_element, tag, attributes ]
             end
           else
             text = @source.read_until("<")
             if text.chomp!("<")
               @source.position -= "<".bytesize
             end
+            if @tags.size.zero? && !@root_tag.nil?
```

Comment:
* Could you reduce method calls?
* Could you use `and` not `&&`?

```suggestion
            if @tags.empty? and !@root_tag.nil?
```

Reply:
OK.
I see.